### DISCUSSION TITLE
Use FakeXrmEasy with Third Party Extended Plugins

### DISCUSSION THREAD

**Discussion by Gmanunta81GFPS** [jordimontana82/fake-xrm-easy#549] [Q&A] [answered]
I currently using FakeXrmEasy with standard Plugins. I works great for me
Now, I would like to integrate with third Party Plugin of which I have the API and Dll to use 

for example --> public class ContactUpdateAddress : PluginBase where PluginBase Extends (IPlugin)
the execution method is -->  protected override void OnExecute(PluginContext pluginContext) where PluginContext extends ContextBase and context base is defined in this way (below the the extration of the class from object metadata) 

What Could I do to integrate FakeXrmEasy with External Third Party Plugin implementation? 

using System;
using Microsoft.Xrm.Sdk;
using ThirdParty.CRM.Configuration;
using ThirdParty.Logging;

namespace ThirdParty.CRM
{
    public abstract class ContextBase
    {
        public ContextBase(ILogSource log);
        public ContextBase(SourceType sourceType, string sourceName);
        protected internal ContextBase();

        public virtual ParameterCollection SharedVariables { get; }
        public virtual ParameterCollection OutputParameters { get; }
        public virtual ParameterCollection InputParameters { get; }
        public abstract IExecutionContext ExecutionContext { get; }
        public abstract Guid UserId { get; }
        public ITracingService TracingService { get; protected set; }
        public IOrganizationServiceFactory OrganizationServiceFactory { get; protected set; }
        public IOrganizationService OrganizationService { get; protected set; }
        public Terna.Localization.Localization Localization { get; protected set; }
        public ILogSource Logger { get; protected set; }
        public Configuration.Configuration Configuration { get; protected set; }
        public IConfigurationProvider ConfigurationProvider { get; protected set; }
        public virtual Guid? TransactionId { get; protected set; }
        public virtual Guid? CorrelationId { get; protected set; }
        public virtual bool IsInTransaction { get; protected set; }
        public ExecutionMode ExecutionMode { get; protected set; }
        public IsolationMode IsolationMode { get; protected set; }

        protected static ExecutionMode GetExecutionModeFromExecutionContext(IExecutionContext executionContext);
        protected static IsolationMode GetIsolationModeFromExecutionContext(IExecutionContext executionContext);
        protected void OnApplyConfigurationToLogger(AdditionalTargetCollection additionalTargets);
        protected virtual void OnLoadConfiguration(IExecutionContext executionContext);
        protected virtual void OnLoadConfigurationProvider();
    }
}

**Comment by BetimBeja** [comment]
Hello @Gmanunta81GFPS ,
you can test the implementation of your plugins which extend the ThirdParty PluginBase with FakeXrmEasy. If you are using constructors with parameters for your plugins you should use the overloads which accept the plugin instance for `ExecutePluginWith` and other methods.
Let me know if I understood your problem correctly.
Betim.

**Comment by Gmanunta81GFPS** [comment]
Hi Betim,

Thanks for the reply.
Looks like something missing

**_FakeXrmEasy.PullRequestException: 'Exception: The specified service type is not supported. This functionality is not available yet. Please consider contributing to the following Git project https://github.com/jordimontana82/fake-xrm-easy by cloning the repository and issuing a pull request.'_**

**_[External Code]
    3party.ServiceProviders.GetService<T>(System.IServiceProvider) in ServiceProviders.cs
    3party.CRM.PluginBase.Execute(System.IServiceProvider) in PluginBase.cs
    [External Code]
    3party.CRM.Customer.Testing.FakeXrmUnitTestProject.PLUGINUnitTest.ContactUpdateaddresstester() in PLUGINUnitTest.cs_**

**Comment by jordimontana82** [comment]
Hi @Gmanunta81GFPS ,

If the custom plugin implements IPlugin it should be possible to use FakeXrmEasy like with any other plugin. We do this to test our own plugins that inherit from a PluginBase class which is what you're asking I suppose.

However, without the plugin source code, is going to be difficult to figure out what's going on there.... you might to follow up with your 3rd party provider?

It seems that 3rd party plugin code is asking for a specific service for which there is no default mock. If you could tell us what that service is it could help.

**Reply by BetimBeja** [reply]
Perhaps we could make the type (service) name part of the not implemented exception message to better understand these cases.

**Reply by jordimontana82** [reply]
Added a new issue to track this as an enhancement: #550

**Comment by Gmanunta81GFPS** [comment]
Hi Jordi, Hi Betim

Thanks for the quick reply (I am one of your supporter)

This is the initial code

  public void ContactUpdateaddresstester()
        {
            var fakedContext = new XrmFakedContext();

           ServicePointManager.SecurityProtocol = (SecurityProtocolType)3072;
            var context = new XrmRealContext
            {
                //ProxyTypesAssembly = typeof(InterestPreCreateDeleteBuildAppointmentInterestTextField).Assembly,
                ConnectionStringName = "CRMOnline"
            };
            var guid1 = Guid.NewGuid();
            var target = new Entity("contact") { Id = guid1 };


            //Execute our plugin against the selected target
            var plugCtx = fakedContext.GetDefaultPluginContext();


            context.ExecutePluginWith<ContactUpdateAddress>(plugCtx);


        }

Second part
    /// <summary>
    /// This Plugin adds the country from the optionset into the address field
    /// </summary>
    [PluginEntity(typeof(Contact)),
    PluginMessage(PluginMessage.Create, PluginMessage.Update),
    PluginStage(PluginStage.PostOperation)]
    public class ContactUpdateAddress : PluginBase
    {
        protected override bool OnCheckFilter(PluginContext pluginContext)
        {
            //prevent infinity loop
            if (pluginContext.Depth > 2)
            {
                return false;
            }
            return base.OnCheckFilter(pluginContext);
        }

        /// <summary>
        /// OnExecute
        /// </summary>
        /// <param name="pluginContext">The plugin context.</param>
        protected override void OnExecute(PluginContext pluginContext)
        { …………………..

Third part

public abstract class PluginBase : IPlugin
    {
        protected PluginBase();
        protected PluginBase(string unsecureConfiguration, string secureConfiguration);

        protected internal string SecureConfiguration { get; }
        protected internal string UnsecureConfiguration { get; }

        protected static void OnValidateRegistration(Type pluginType, PluginContext pluginContext);
        protected internal static void OnValidateRegistration(Type pluginType, PluginContext pluginContext, bool inheritAttributes);
        public void Execute(PluginContext pluginContext);
        public void Execute(IServiceProvider serviceProvider);
        protected virtual Guid? GetOrganizationServiceUserId(IPluginExecutionContext pluginExecutionContext);
        protected virtual bool OnCheckFilter(PluginContext pluginContext);
        protected virtual bool OnCheckFilterPreValidation(IPluginExecutionContext pluginExecutionContext);
        protected virtual bool OnCheckMessageContextPreValidation(IPluginExecutionContext pluginExecutionContext);
        protected abstract void OnExecute(PluginContext pluginContext);
        protected virtual void OnExecutionCompleted(PluginContext pluginContext);
        protected virtual void OnPluginContextCreated(PluginContext pluginContext);
        protected virtual void OnValidateRegistration(PluginContext pluginContext);
    }

I hope this help

**Reply by BetimBeja** [reply]
You should check your ServiceProviders.cs
```
3party.ServiceProviders.GetService<T>(System.IServiceProvider) in ServiceProviders.cs
```
What types are specified in the calls to `3party.ServiceProviders.GetService<T>(System.IServiceProvider)`? Only a few services are available in plugins AFAIK.

**Reply by jordimontana82** [reply]
In addition to that, it seems you're executing a plugin using the XrmRealContext instance not the XrmFakedContext?? That's not a envisioned scenario to use the ExecutePluginWith method ... @Gmanunta81GFPS

**Comment by Gmanunta81GFPS** [comment]
Oaky

What should I do ? with method should I use ?

**Comment by Gmanunta81GFPS** [comment, marked as answer]
Hi everyone,

Thank you very much for your support Betim.

the issue is in this line of code 

serviceProvider.GetService<IPluginContextFactory>();

the interface IPluginContextFactory is not implemented in the Standard IServiceProvider serviceProvider object and the system is going to catch the exception (unfortunately for me )

**Reply by Gmanunta81GFPS** [reply]
@BetimBeja let me know if I missed something

**Reply by BetimBeja** [reply]
That is correct. As documented [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.xrm.sdk.iplugin.execute?view=dynamics-general-ce-9#Microsoft_Xrm_Sdk_IPlugin_Execute_System_IServiceProvider_) the `IPluginContextFactory` is not part of the services you can request from the `IServiceProvider`
